### PR TITLE
JDBC instrumentation

### LINKED ISSUE #35: Zero garbage JDBC tracing
P6spy allocates objects on each query:

https://github.com/p6spy/p6spy/blob/f9f852e895419a8adcda590618aec4ae47513cfe/src/main/java/com/p6spy/engine/wrapper/PreparedStatementWrapper.java#L59-L64

So investigate if p6spy could be easily modified to allow for object pooling or if we should rather implement our own JDBC wrapper. P6spy has an extensive test suite for all sorts of databases and application servers so it would be nice to leverage this.

### PR DESCRIPTION
**Pull request by felixbarny**
closes elastic/apm-agent-java#35

### CODE REVIEW

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/jdbc/StatementInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    @Override
+    public void init(ElasticApmTracer tracer) {
+        StatementInstrumentation.tracer = tracer;
+        StatementInstrumentation.jdbcEventListener = new ApmJdbcEventListener(tracer);
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return not(isInterface())
+            .and(nameContains("Statement"))
```
Comment (tylerbenson) [inline apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/jdbc/StatementInstrumentation.java]: Just out of curiosity, why was this needed?  Is this a performance optimization to avoid the `isSubTypeOf` check?

Comment (felixbarny) [reply apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/jdbc/StatementInstrumentation.java]: See the comment in https://github.com/elastic/apm-agent-java/blob/master/apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/servlet/ServletInstrumentation.java#L69
Basically it’s because the isSubType marcher can be pretty expensive. It causes a lot of type descriptions to be parsed by ByteBuddy. I stagemonitor, I cached TypeDescriptions to improve instrumentation performance (startup times). A user experienced an OOME because of an isSubType matcher. But I should re evaluate with a recent Byte Buddy version.

**Comment by felixbarny** [comment]
TODO: more integration tests
